### PR TITLE
Add constructor with `RenderSettings` to `ComposePanel`

### LINKED ISSUE #4870: Being able to create a ComposePanel with vsync disabled
Hi,
I've been discussing the current latency issue with @igordmn , especially when working with inputs. One of the possible solutions to reduce latency is to disable vsync, but it's not a good option because it will make the program slower. 
So Igor suggested having 2 panels one for ui (with vsync) and one for drawing (without vsync).
We will need the ability to configure `ComposePanel(skiaLayerProperties = SkiaLayerProperties(isVsyncEnabled = false))`

Thanks!

### PR DESCRIPTION
**Pull request by elijah-semyonov**
Fixes https://github.com/JetBrains/compose-multiplatform/issues/4870

### Features - Desktop
- Add constructor with `RenderSettings` to `ComposePanel`. Added a class `RenderSettings` with  `val isVsyncEnabled: Boolean?`. When set to `true` gives a hint to renderer implementation of the particular `ComposePanel` to reduce the latency between the input and visual changes in exchange for possible screen tearing.

### CODE REVIEW

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposePanel.desktop.kt [hunk 1]
```diff
@@ -46,8 +47,29 @@ import org.jetbrains.skiko.SkiaLayerAnalytics
...
+     * This flag is a hint and may not be respected by the underlying implementation.
+     * The reduction of latency may come at the cost of visual artifacts like screen tearing.
+     *
+     * If not set, the parameter from [SkikoProperties] will be used.
+     * @see [SkikoProperties.vsyncEnabled]
+     *
+     * @param skiaLayerAnalytics Analytics that helps to know more about SkiaLayer behaviour.
+     */
+    constructor(
+        reducePresentationLatency: Boolean,
```
Comment (MatkovIvan) [inline compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposePanel.desktop.kt]: Let's add it as a last parameter to main constructor (marked as `@ExperimentalComposeUiApi`)

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposePanel.desktop.kt [hunk 2]
```diff
@@ -46,8 +47,29 @@ import org.jetbrains.skiko.SkiaLayerAnalytics
 class ComposePanel @ExperimentalComposeUiApi constructor(
     private val skiaLayerAnalytics: SkiaLayerAnalytics,
 ) : JLayeredPane() {
+    private var reducePresentationLatency: Boolean? = null
```
Comment (MatkovIvan) [inline compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposePanel.desktop.kt]: Why not just `vsyncEnabled` to have 1-1 match with skiko property that it's actually set?

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposePanel.desktop.kt [hunk 3]
```diff
@@ -46,8 +47,29 @@ import org.jetbrains.skiko.SkiaLayerAnalytics
...
+
     constructor() : this(SkiaLayerAnalytics.Empty)
 
+    /**
+     * Creates a ComposePanel with the specified forceReducedLatency flag.
```
Comment (MatkovIvan) [inline compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposePanel.desktop.kt]: I don't see any flags with `forceReducedLatency` name

**Comment by m-sasha** [comment]
While I don't think this change is bad in and of itself (it's good to give more granularity to the vsync flag), I don't like how this is being used as a workaround for the presentation latency issue. It solves a very small aspect of the problem, and in a very inconvenient way.

- It doesn't solve the issue for fully Compose (no Swing) apps. 
- It doesn't solve the issue for regular UI, where tearing looks bad.

I'd much rather we solved the underlying issue; if Swing/AWT can do it, why not us?

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposePanel.desktop.kt (hunk 2 above)
Comment (m-sasha) [inline compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposePanel.desktop.kt]: Agree with Ivan that `vsyncEnabled` is better. `reducePresentationLatency` doesn't hint at any downsides - why would anyone *not* want to reduce presentation latency?

But maybe consider going even farther and give it a (fake) enum, e.g. `PresentationMode.VSYNC`, `PresentationMode.IMMEDIATE` etc, if there's a chance we could have more modes...

**Comment by igordmn** [comment]
> I'd much rather we solved the underlying issue; if Swing/AWT can do it, why not us?

I would too, but as we discussed, it is much harder, and I am not sure we are able to solve it in the foreseeble future. In the meantime, we can provide a property, that even makes sense outside of this issue.

**Comment by igordmn** [comment]
> if Swing/AWT can do it, why not us?

If we just disable vsync, we will have:
- tearing (depends on the system, the system window compositor can add additional synchronization)
- animation issues
  - frame drops
  - inconsistent animation frame times (animations won't be smooth)
  - unlimited FPS (consumes CPU)
- too early recomposition, that results the frame after the next to drop

Swing can have the same issues.

**Comment by m-sasha** [comment]
> If we just disable vsync, we will have:

Are there no other solutions?

> Swing can have the same issues.

But in practice, it doesn't. Swing has low-latency presentation and no tearing out-of-the-box.

**Comment by igordmn** [comment]
> Are there no other solutions?

This is the nature of vsync. If we sync with it, we have 1-frame latency. We sync with it to solve the issues I mentioned.

Animations should always be in sync with v-sync, to avoid them. The solution might be to move input reaction changes out of vsync, but it can be not trivial. We might even need a new API that user should call explicitly, not sure.

**Comment by igordmn** [comment]
We also might have unnecessary frame lags in other places. Most probably it is easier to fix them than the vsync latency. But we need to profile the issue to understand if we have them.

**Comment by m-sasha** [comment]
> This is the nature of vsync.

But, again, why is this problem not present in Swing, and (I assume) native frameworks?

Can we double buffer ourselves to avoid tearing?

> We also might have unnecessary frame lags in other places. 

The problem reproduces in pure Skiko, and it appears that most (possibly all) the delay comes from vsync.

**Comment by igordmn** [comment]
> But, again, why is this problem not present in Swing, and (I assume) native frameworks?

Besides tearing I mentioned the animation issue, which Swing didn't solve. It doesn't sync it with vsync. I haven't investigated it properly in the past though. I discovered this when I tried to implement smooth animations on Swing, and didn't find a way.

I am not saying that there is no proper way, I am saying that a proper way needs time to investigate/design/implement. And I am all for doing that.

> The problem reproduces in pure Skiko

What I meant that we can have issues in Compose, and solving them might also help. Skiko also can have additional lags besides that caused by vsync.

**Comment by igordmn** [comment]
> But, again, why is this problem not present in Swing, and (I assume) native frameworks?

I would look at the Android sources (`Choreographer`). From my experience, it solves the latency and the animation issues. As I remember, it was introduced as Project Butter.

**Comment by m-sasha** [comment]
> I am not saying that there is no proper way, I am saying that a proper way needs time to investigate/design/implement. And I am all for doing that.

Ok, understood.

> What I meant that we can have issues in Compose, and solving them might also help. Skiko also can have additional lags besides that caused by vsync.

The experiments I've done show that this isn't the case. Or at least the contribution of other factors (in Skiko or Compose) to the latency are minimal.
1. The latency in a pure Skiko reproducer is very similar to the latency of a Compose reproducer.
2. The latency in pure Skiko with vsync turned off is negligible/acceptable - probably just the remaining, expected 1-frame delay.

**Comment by m-sasha** [comment]
Note that merely setting
```
device.layer.displaySyncEnabled = NO;
```
while keeping everything else related to vsync (e.g. calling `DisplayLinkThrottler.waitVSync`) also causes the problem to go away (but causes tearing).

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposePanel.desktop.kt (hunk 2 above)
Comment (elijah-semyonov) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposePanel.desktop.kt]: Because it's a bit misleading. We still perform frame-rate(vsync-based) throttling (at least on macOS).
This change in fact affects only vsync property on CAMetalLayer (which is exclusive to macOS).

**Comment by elijah-semyonov** [comment]
> Note that merely setting
> 
> ```
> device.layer.displaySyncEnabled = NO;
> ```
> 
> while keeping everything else related to vsync (e.g. calling `DisplayLinkThrottler.waitVSync`) also causes the problem to go away (but causes tearing).

Yes, that's true. The introduced flag doesn't affect throttler behavior.

**Comment by elijah-semyonov** [comment]
Thank you for feedback and discussion, I'll do deeper investigation of other graphics backends and see if I can define a common denominator here.

**Comment by MohamedRejeb** [comment]
I noticed that even with vsync disabled, you can get better results with requesting redraw only 1 time per frame.
I'm on 60 fps so the frame duration is around 16ms. Calling `skiaLayer.needRedraw()` on each skiko `onRender()` call results calling `onRender()` each 8ms (2 times per frame). I did it manually and I called `skiaLayer.needRedraw()` each 16 ms and the results was better. 
I'm not sure if there's a way to prevent calling draw method more than one time on each frame even before reaching the native calls, maybe it can be done on the `FrameDispatcher`.

**Comment by elijah-semyonov** [comment]
> I noticed that even with vsync disabled, you can get better results with requesting redraw only 1 time per frame. I'm on 60 fps so the frame duration is around 16ms. Calling `skiaLayer.needRedraw()` on each skiko `onRender()` call results calling `onRender()` each 8ms (2 times per frame). I did it manually and I called `skiaLayer.needRedraw()` each 16 ms and the results was better. I'm not sure if there's a way to prevent calling draw method more than one time on each frame even before reaching the native calls, maybe it can be done on the `FrameDispatcher`.

Which platform did you test it on? 
It doesn't seem a valid behavior on macOS. We already use framerate throttling disregarding the vsyncEnabled (which control the presentation, not scheduling)

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/scene/skia/SwingSkiaLayerComponent.desktop.kt [hunk 4]
```diff
@@ -66,7 +66,7 @@ internal class SwingSkiaLayerComponent(
                 mediator.preferredSize
             }
 
-            override fun getAccessibleContext(): AccessibleContext? {
+            override fun getAccessibleContext(): AccessibleContext {
```
Comment (MatkovIvan) [inline compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/scene/skia/SwingSkiaLayerComponent.desktop.kt]: Please revert unrelated changes

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposePanel.desktop.kt [hunk 5]
```diff
@@ -33,8 +33,47 @@ import java.awt.event.FocusListener
...
+ * @property adapterPriority Specifies the GPU that will be selected for rendering.
+ * Default value is [SkikoProperties.gpuPriority].
+ */
+
+@ExperimentalComposeUiApi
+class RendererConfiguration(
+    val isVsyncEnabled: Boolean = SkikoProperties.vsyncEnabled,
+    val isVsyncFramelimitFallbackEnabled: Boolean = SkikoProperties.vsyncFramelimitFallbackEnabled,
+    val renderApi: GraphicsApi = SkikoProperties.renderApi,
+    val adapterPriority: GpuPriority = SkikoProperties.gpuPriority
```
Comment (MatkovIvan) [inline compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposePanel.desktop.kt]: We should not introduce skiko types as part of Compose public API

**Review by MatkovIvan — CHANGES_REQUESTED** [review]

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/scene/skia/SwingSkiaLayerComponent.desktop.kt (hunk 4 above)
Comment (elijah-semyonov) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/scene/skia/SwingSkiaLayerComponent.desktop.kt:69]: It's a compilation error without this one on my machine

**Comment by elijah-semyonov** [comment]
There is quite a lot of effects for fully exposing `SkikoLayerProperties`, so @igordmn and me decided to start small and fix the immediate issue with latency on macOS being a blocker for some users.

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposePanel.desktop.kt [hunk 6]
```diff
@@ -46,8 +46,26 @@ import org.jetbrains.skiko.SkiaLayerAnalytics
...
+     *
+     * @param skiaLayerAnalytics Analytics that helps to know more about SkiaLayer behaviour.
+     * SkiaLayer is underlying class used internally to draw Compose content.
+     * Implementation usually uses third-party solution to send info to some centralized analytics gatherer.
+     * @param renderingSettings Configuration class for rendering settings.
+     */
+    @ExperimentalComposeUiApi
+    constructor(
+        skiaLayerAnalytics: SkiaLayerAnalytics = SkiaLayerAnalytics.Empty,
+        renderingSettings: RenderingSettings,
```
Comment (MatkovIvan) [inline compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposePanel.desktop.kt]: the main constructor is experimental, you can just add a new param to it

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/RenderingSettings.desktop.kt [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+ * When null, the internal implementation will use the global configuration from [SkikoProperties.vsyncEnabled]
+ *
+ * This flag has no effect if [ComposeFeatureFlags.useSwingGraphics] is true. In this case Compose
+ * will render the content to Swing provided offscreen buffer and the presentation will be controlled
+ * by Swing.
+ */
+@ExperimentalComposeUiApi
+class RenderingSettings(
+    val isVsyncEnabled: Boolean?
+
```
Comment (MatkovIvan) [inline compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/RenderingSettings.desktop.kt]:
```suggestion
```

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/RenderingSettings.desktop.kt [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+ * When false, the internal implementation will not attempt to synchronize drawable presentations
+ * it can reduce latency but may introduce visual artifacts like screen tearing.
+ * When null, the internal implementation will use the global configuration from [SkikoProperties.vsyncEnabled]
+ *
+ * This flag has no effect if [ComposeFeatureFlags.useSwingGraphics] is true. In this case Compose
+ * will render the content to Swing provided offscreen buffer and the presentation will be controlled
+ * by Swing.
+ */
+@ExperimentalComposeUiApi
+class RenderingSettings(
```
Comment (MatkovIvan) [inline compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/RenderingSettings.desktop.kt]: Should we add equals/hashCode to it?

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/scene/ComposeContainer.desktop.kt [hunk 9]
```diff
@@ -78,6 +79,7 @@ import org.jetbrains.skiko.SkiaLayerAnalytics
  *  for window coordinate space.
  * @property useSwingGraphics Flag indicating if offscreen rendering to Swing graphics is used.
  * @property layerType The type of layer used for Popup/Dialog.
+ * @property renderingSettings The settings for rendering.
```
Comment (MatkovIvan) [inline compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/scene/ComposeContainer.desktop.kt]: Add a note that it will work only with `useSwingGraphics == true`

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/scene/ComposeContainer.desktop.kt [hunk 10]
```diff
@@ -338,7 +341,13 @@ internal class ComposeContainer(
         return if (useSwingGraphics) {
             SwingSkiaLayerComponent(mediator, renderDelegate, skiaLayerAnalytics)
         } else {
-            WindowSkiaLayerComponent(mediator, windowContext, renderDelegate, skiaLayerAnalytics)
+            WindowSkiaLayerComponent(
+                mediator,
```
Comment (MatkovIvan) [inline compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/scene/ComposeContainer.desktop.kt:345]: Add explicit parameter names if it's not one-liner

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/scene/skia/WindowSkiaLayerComponent.desktop.kt [hunk 11]
```diff
@@ -47,6 +50,13 @@ internal class WindowSkiaLayerComponent(
             // apply `checkNotNull` for "non-null" field.
             checkNotNull(mediator.accessible)
         },
+        properties = run {
+            val defaultProperties = SkiaLayerProperties()
+
+            SkiaLayerProperties(
+                isVsyncEnabled = renderingSettings.isVsyncEnabled ?: defaultProperties.isVsyncEnabled,
+            )
+        },
```
Comment (MatkovIvan) [inline compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/scene/skia/WindowSkiaLayerComponent.desktop.kt]:
```suggestion
        properties = SkiaLayerProperties(
            isVsyncEnabled = renderingSettings.isVsyncEnabled ?: SkiaLayerProperties.Default.isVsyncEnabled,
        ),
```

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/RenderingSettings.desktop.kt (hunk 8 above)
Comment (elijah-semyonov) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/RenderingSettings.desktop.kt]: Nop, we don't do any comparisons/caching/using it as a key, etc.

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/scene/ComposeContainer.desktop.kt (hunk 9 above)
Comment (elijah-semyonov) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/scene/ComposeContainer.desktop.kt]: It's in the doc of RenderSettings.

Comment (elijah-semyonov) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/scene/ComposeContainer.desktop.kt]: In general it's not true, because we can add settings, that _will_ work with swing

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/scene/ComposeContainer.desktop.kt (hunk 10 above)
Comment (elijah-semyonov) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/scene/ComposeContainer.desktop.kt:345]: Do we add them when variables match parameter names or are obvious?

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/scene/skia/WindowSkiaLayerComponent.desktop.kt (hunk 11 above)
Comment (elijah-semyonov) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/scene/skia/WindowSkiaLayerComponent.desktop.kt]: There can be multiple properties in future, this approach will reconstruct SkiaLayerProperties every single time. There is no `Default` in `SkiaLayerProperties.Companion`

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/scene/ComposeContainer.desktop.kt (hunk 10 above)
Comment (MatkovIvan) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/scene/ComposeContainer.desktop.kt:345]: It's better to add it to be on the safe side during possible refactoring

**Review by igordmn — APPROVED** [review]

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposePanel.desktop.kt [hunk 12]
```diff
@@ -42,11 +42,13 @@ import org.jetbrains.skiko.SkiaLayerAnalytics
  * @param skiaLayerAnalytics Analytics that helps to know more about SkiaLayer behaviour.
  * SkiaLayer is underlying class used internally to draw Compose content.
  * Implementation usually uses third-party solution to send info to some centralized analytics gatherer.
+ * @param renderingSettings Configuration class for rendering settings.
  */
 class ComposePanel @ExperimentalComposeUiApi constructor(
     private val skiaLayerAnalytics: SkiaLayerAnalytics,
+    private val renderingSettings: RenderingSettings
```
Comment (MatkovIvan) [inline compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/ComposePanel.desktop.kt]:
```suggestion
    private val renderingSettings: RenderingSettings = RenderingSettings.Default
```

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/RenderingSettings.desktop.kt [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+    val isVsyncEnabled: Boolean?
```
Comment (m-sasha) [inline compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/RenderingSettings.desktop.kt:40]: It's better/cleaner to put this in a (fake) enum, e.g. `VSyncMode.Enabled`, `VSyncMode.Disabled`, and `VsyncMode.Default`. Although maybe it's overkill here; not sure.

**Review by m-sasha — APPROVED** [review]

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/RenderingSettings.desktop.kt [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+ * When false, the internal implementation will not attempt to synchronize drawable presentations
+ * it can reduce latency but may introduce visual artifacts like screen tearing.
+ * When null, the internal implementation will use the global configuration from [SkikoProperties.vsyncEnabled]
+ *
+ * This flag has no effect if [ComposeFeatureFlags.useSwingGraphics] is true. In this case Compose
+ * will render the content to Swing provided offscreen buffer and the presentation will be controlled
+ * by Swing.
+ */
+@ExperimentalComposeUiApi
+class RenderingSettings(
```
Comment (m-sasha) [inline compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/RenderingSettings.desktop.kt]: Maybe `RenderSettings` instead? A bit shorter and doesn't lose any meaning.

Comment (elijah-semyonov) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/RenderingSettings.desktop.kt]: Yeah, sounds fine

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/RenderingSettings.desktop.kt (hunk 13 above)
Comment (elijah-semyonov) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/RenderingSettings.desktop.kt:40]: I'd leave it as it is now (ExperimentalAPI after all), I haven't yet performed an investigation about different presentation modes on non-Metal rendering backends.

**Comment by MohamedRejeb** [comment]
Btw the latency issue exists on windows as well (didn't try linux) and disabling vsync reduces the latency significantly there. So it's not only a macos thing.

**Review by MatkovIvan — APPROVED** [review]

**Comment by igordmn** [comment]
@elijah-semyonov, please actualise Release Notes
